src/allocation_engine/interval_tree.rs

```diff
@@ -353,11 +355,106 @@ impl InnerNode {
         self.height = max(height(&self.left), height(&self.right)) + 1;
     }
 
-    /// Updates the sub-tree to keep balance.
+    /// Insert a new node in the subtree. After the node is inserted the
+    /// tree will be balanced. The node_state parameter is needed because in
+    /// the AddressAllocator allocation logic we will need to insert both free
+    /// and allocated nodes.
     #[allow(dead_code)]
-    fn update_node(mut self: Box<Self>) -> Box<Self> {
+    pub(crate) fn insert(
+        mut self: Box<Self>,
+        key: Range,
+        node_state: NodeState,
+    ) -> Result<Box<Self>> {
+        // The InnerNode structure has 48 a length of 48 bytes. With other nested
+        // calls that are made during the insertion process the size occupied
+        // on the stack by just one insert call is around 122 bytes. Considering
+        // that the default stack size on Linux is 8K we could make around 73
+        // calls to insert method before confronting with an stack overflow. To
+        // be cautious we will use 50 as the maximum height of the tree. A
+        // maximum height of 50 will result in the possibility to allocate
+        // (2^50 - 1) memory slots. Considering the imposed maximum height the
+        // recursion is safe to use.
+        if (self.height + 1) > 50 {
+            return Err(Error::Overflow);
+        }
+        if self.key.overlaps(&key) {
+            return Err(Error::Overlap(key, self.key));
+        }
+        match self.key.cmp(&key) {
+            // It is not possible for a range to be equal with an existing node
+            // as the overlaps method will also catch this case and return the
+            // corresponding error code.
+            Ordering::Equal => unreachable!(),
+            Ordering::Less => match self.right {
+                None => self.right = Some(Box::new(InnerNode::new(key, node_state))),
+                Some(right) => {
+                    self.right = Some(right.insert(key, node_state)?);
+                }
+            },
+            Ordering::Greater => match self.left {
+                None => self.left = Some(Box::new(InnerNode::new(key, node_state))),
+                Some(left) => {
+                    self.left = Some(left.insert(key, node_state)?);
+                }
+            },
+        }
         self.update_cached_height();
-        self.rotate()
+        Ok(self.rotate())
+    }
+
+    /// Update the state of an old node. This method should be used when we
+    /// find an existing node with the state `NodeState::Free` that satisfies
+    /// all constraints of an allocation request. The recursion is safe as we
+    /// have in place a maximum height for the tree.
```

Comment:
We shouldn't have comments about how things are implemented in the doc comments. This should instead be written as a normal comment, using `//`.